src/ZulipMobile.js

```diff
@@ -31,7 +31,12 @@ export default (): React$Node => (
             <ThemeProvider>
               <InitialNavigationDispatcher>
                 <BackNavigationHandler>
-                  <AppWithNavigation ref={NavigationService.reduxContainerRef} />
+                  <AppContainer
+                    // `static navigationOptions` and `static router` not
+                    // being handled properly
+                    // $FlowFixMe
+                    ref={NavigationService.appContainerRef}
```

Comment:
Hmm interesting. Here's the error message I get when I remove the fixme:
```
Cannot create AppContainer element because:
 • Either mixed [1] is incompatible with React.ComponentType [2] in property
   current of property ref.
 • Or a call signature declaring the expected parameter / return type is missing
   in object type [3] but exists in function type [4] in property ref.
 • Or withRouter [5] is not a React component.
 • Or withOptionalNavigationOptions [6] is not a React component.

     src/ZulipMobile.js
      29│           <TranslationProvider>
      30│             <ThemeProvider>
      31│               <InitialNavigationDispatcher>
      32│                 <AppContainer
      33│                   // `static navigationOptions` and `static router` not
      34│                   // being handled properly
      35│                   ref={NavigationService.appContainerRef}
      36│                 />
      37│               </InitialNavigationDispatcher>
      38│             </ThemeProvider>
      39│           </TranslationProvider>

     flow-typed/npm/react-navigation_v4.x.x.js
 [2] 413│   > = React$ComponentType<Props> &
 [5] 414│     withRouter<State, Options> &
 [6] 415│     withOptionalNavigationOptions<Options>;

     /tmp/flow/flowlib_1bc53cbd/react.js
 [1] 159│ declare type React$ComponentType<-Config> = React$AbstractComponent<Config, mixed>;
        :
 [4] 197│   | ((React$ElementRef<ElementType> | null) => mixed)
        :
 [3] 246│   ): {|current: null | T|};
```

I tried removing those statics from the `NavigationContainer` definition in the libdef:
```diff
--- a/flow-typed/npm/react-navigation_v4.x.x.js
+++ b/flow-typed/npm/react-navigation_v4.x.x.js
@@ -410,9 +410,9 @@ declare module 'react-navigation' {
     State: NavigationState,
     Options: {...},
     Props: NavigationContainerProps<Options, State>,
-  > = React$ComponentType<Props> &
+  > = React$ComponentType<Props> /* &
     withRouter<State, Options> &
-    withOptionalNavigationOptions<Options>;
+    withOptionalNavigationOptions<Options> */;
```
That eliminated the last two of those four errors, but the first two remain. So I think the issue is more basic.

... OK, on some experimentation, I think I see the cause! It looks like the type parameter to `React.createRef` should be a `React$ElementRef` type:
```diff
 const appContainerRef = React.createRef<
-  NavigationContainer<NavigationState, { ... }, NavigationContainerProps<{ ... }, NavigationState>>,
+  React$ElementRef<
+    NavigationContainer<NavigationState, { ... }, NavigationContainerProps<{ ... }, NavigationState>>,
+  >,
 >();
```
That resolves this fixme, and also the one on `appContainerRef.current.dispatch`.

I got there by looking at the type definitions mentioned in that Flow error message, and in particular looking closely at the `createRef` declaration:
```js
  declare export function createRef<T>(
  ): {|current: null | T|};
```
and comparing with this one in the same file (and also mentioned in the error message):
```js
/**
 * The type of the ref prop available on all React components.
 */
declare type React$Ref<ElementType: React$ElementType> =
  | { -current: React$ElementRef<ElementType> | null, ... }
  | ((React$ElementRef<ElementType> | null) => mixed)
  | number | string;
```
Note the types of the `current` property.

I'll push a commit with that change to the tip of this PR branch.

In the course of finding that, I also discovered some small bugs in the libdef and other small ways it can be improved. I pushed those fixes to [another branch](https://github.com/zulip/zulip-mobile/compare/master...gnprice:dev-nav-libdef), though they turn out not to be necessary here.